homeassistant/components/waze_travel_time/__init__.py

```diff
@@ -1,24 +1,182 @@
+)
+from homeassistant.helpers.httpx_client import get_async_client
+from homeassistant.helpers.selector import (
+    BooleanSelector,
+    SelectSelector,
+    SelectSelectorConfig,
+    SelectSelectorMode,
+    TextSelector,
+)
 
-from .const import DOMAIN, SEMAPHORE
+from .const import (
+    CONF_AVOID_FERRIES,
+    CONF_AVOID_SUBSCRIPTION_ROADS,
+    CONF_AVOID_TOLL_ROADS,
+    CONF_DESTINATION,
+    CONF_ORIGIN,
+    CONF_REALTIME,
+    CONF_UNITS,
+    CONF_VEHICLE_TYPE,
+    DEFAULT_VEHICLE_TYPE,
+    DOMAIN,
+    METRIC_UNITS,
+    REGIONS,
+    SEMAPHORE,
+    UNITS,
+    VEHICLE_TYPES,
+)
 
 PLATFORMS = [Platform.SENSOR]
 
+SERVICE_GET_TRAVEL_TIMES = "get_travel_times"
+SERVICE_GET_TRAVEL_TIMES_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_ORIGIN): TextSelector(),
+        vol.Required(CONF_DESTINATION): TextSelector(),
+        vol.Required(CONF_REGION): SelectSelector(
+            SelectSelectorConfig(
+                options=sorted(REGIONS),
+                mode=SelectSelectorMode.DROPDOWN,
+                translation_key=CONF_REGION,
+            )
+        ),
+        vol.Optional(CONF_REALTIME, default=False): BooleanSelector(),
+        vol.Optional(CONF_VEHICLE_TYPE, default=DEFAULT_VEHICLE_TYPE): SelectSelector(
+            SelectSelectorConfig(
+                options=sorted(VEHICLE_TYPES),
+                mode=SelectSelectorMode.DROPDOWN,
+                translation_key=CONF_VEHICLE_TYPE,
+            )
+        ),
+        vol.Optional(CONF_UNITS, default=METRIC_UNITS): SelectSelector(
+            SelectSelectorConfig(
+                options=sorted(UNITS),
+                mode=SelectSelectorMode.DROPDOWN,
+                translation_key=CONF_UNITS,
+            )
+        ),
+        vol.Optional(CONF_AVOID_TOLL_ROADS, default=False): BooleanSelector(),
+        vol.Optional(CONF_AVOID_SUBSCRIPTION_ROADS, default=False): BooleanSelector(),
+        vol.Optional(CONF_AVOID_FERRIES, default=False): BooleanSelector(),
+    }
+)
+
+_LOGGER = logging.getLogger(__name__)
+
 
 async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
     """Load the saved entities."""
     if SEMAPHORE not in hass.data.setdefault(DOMAIN, {}):
         hass.data.setdefault(DOMAIN, {})[SEMAPHORE] = asyncio.Semaphore(1)
     await hass.config_entries.async_forward_entry_setups(config_entry, PLATFORMS)
+
+    async def async_get_travel_times_service(service: ServiceCall) -> ServiceResponse:
+        httpx_client = get_async_client(hass)
+        client = WazeRouteCalculator(
+            region=service.data[CONF_REGION].upper(), client=httpx_client
+        )
+        response = await async_get_travel_times(
+            client=client,
+            origin=service.data[CONF_ORIGIN],
+            destination=service.data[CONF_DESTINATION],
+            vehicle_type=""
+            if service.data[CONF_VEHICLE_TYPE].upper() == "CAR"
+            else service.data[CONF_VEHICLE_TYPE].upper(),
```

Comment:
I think it would be more logical if this logic is moved to `async_get_travel_times`